client/extensions/woocommerce/app/store-stats/referrers/index.js

```diff
@@ -21,61 +21,165 @@ import StoreStatsPeriodNav from 'woocommerce/app/store-stats/store-stats-period-
-							<tr key={ d.referrer }>
-								<td>{ d.date }</td>
-								<td>{ d.referrer }</td>
-								<td>{ d.product_views }</td>
-								<td>{ d.add_to_carts }</td>
-								<td>{ d.product_purchases }</td>
-								<td>${ d.sales }</td>
-							</tr>
-						) ) }
-					</tbody>
-				</table>
-			</Module>
-			<JetpackColophon />
-		</Main>
-	);
-};
+class Referrers extends Component {
+	static propTypes = {
+		siteId: PropTypes.number,
+		query: PropTypes.object.isRequired,
+		data: PropTypes.array.isRequired,
+		selectedDate: PropTypes.string,
+		unit: PropTypes.oneOf( [ 'day', 'week', 'month', 'year' ] ),
+		slug: PropTypes.string,
+		limit: PropTypes.number,
+	};
+
+	state = {
+		filter: '',
+	};
+
+	onSearch = str => {
+		this.setData( this.props, str );
+	};
+
+	getFilteredSortedData = ( filter, { data, selectedDate, unit } ) => {
+		const unitSelectedDate = getUnitPeriod( selectedDate, unit );
+		const selectedData = find( data, d => d.date === unitSelectedDate ) || { data: [] };
+		const filteredData = filter
+			? selectedData.data.filter( d => d.referrer.match( filter ) )
+			: selectedData.data;
+		return {
+			filteredSortedData: sortBySales( filteredData ),
+			unfilteredDataLength: selectedData.data.length,
+		};
+	};
+
+	getSelectedReferrer = ( filteredSortedData, { queryParams } ) => {
+		let selectedReferrerIndex = null;
+		const selectedReferrer = find( filteredSortedData, ( d, idx ) => {
+			if ( queryParams.referrer && queryParams.referrer === d.referrer ) {
+				selectedReferrerIndex = idx;
+				return true;
+			}
+			return false;
+		} );
+		return {
+			selectedReferrer,
+			selectedReferrerIndex,
+		};
+	};
+
+	setData( props, _filter ) {
+		const isEmpty = _filter === '';
+		const filter = isEmpty ? '' : _filter || this.state.filter;
+		const { filteredSortedData, unfilteredDataLength } = this.getFilteredSortedData(
+			filter,
+			props
+		);
+		const { selectedReferrer, selectedReferrerIndex } = this.getSelectedReferrer(
+			filteredSortedData,
+			props
+		);
+		this.setState( {
+			filter,
+			filteredSortedData,
+			unfilteredDataLength,
+			selectedReferrer,
+			selectedReferrerIndex,
+		} );
+	}
 
-Referrers.propTypes = {
-	siteId: PropTypes.number,
-	query: PropTypes.object.isRequired,
-	data: PropTypes.array.isRequired,
-	selectedDate: PropTypes.string,
-	unit: PropTypes.oneOf( [ 'day', 'week', 'month', 'year' ] ),
-	slug: PropTypes.string,
-};
+	componentWillReceiveProps( nextProps ) {
+		this.setData( nextProps );
+	}
+
+	componentWillMount() {
+		this.setData( this.props );
+	}
+
+	render() {
+		const { siteId, query, selectedDate, unit, slug, translate, queryParams } = this.props;
+		const {
+			filter,
+			filteredSortedData,
+			unfilteredDataLength,
+			selectedReferrer,
+			selectedReferrerIndex,
+		} = this.state;
+		const unitSelectedDate = getUnitPeriod( selectedDate, unit );
+		const showSearch = unfilteredDataLength > LIMIT;
+		const title = `${ translate( 'Store Referrers' ) }${
```

Comment:
Do you think we should use a [placeholder](https://github.com/Automattic/i18n-calypso#mixing-strings-and-markup) here in the translate call for the `queryParams.referrer`?

Reply:
I'm not following exactly. The `queryParams.referrer` shouldn't get translated and there's no markup mixed with strings. Can you explain? Cheers